tech_reports/LLMs/llms.md

```diff
@@ -86,7 +87,231 @@ Authors:
+
+> **Note:** If you see errors whilst running tracy, try this device-only profiling process instead: run with `TT_METAL_DEVICE_PROFILER=1 pytest path/to/test.py` and after the run completes run `tt_metal/tools/profiler/process_ops_logs.py --date` to generate the CSV file.
+
+This CSV file contains a wealth of information recorded from all the device during program execution. To summarize it into a more human-readable format we run the `perf_report.py` tool:
+
+```bash
+python models/perf/perf_report.py OPS_CSV_FILE
+```
+
+The [documentation for this tool](https://github.com/tenstorrent/tt-metal/tree/main/models/perf) describes how to use it to select specific ranges of operations. For device performance we strongly recommend looking at a single layer. You can do this by using `--id-range` or by changing your test to run only a single layer of the model (recommended).
+
+##### What makes a good performance test?
+
+Ideally you should run your model in as close to end-user form as possible, whilst simplifying as much as possible. In practice this means:
+
+* Use tracing (if you are using tracing in production)
+* Skip the first compilation iteration - this adds a lot of one-time host overhead between ops
+* Run a single layer of the model - but be aware of which ops will be run for every single layer and which ones are only run at the start and end (e.g. embedding, final norm and LM head)
+* Add a tracy signpost e.g. `tracy.signpost("Performance pass")` just before the part of your run you want to record - this will be focused on by default by `perf_report.py`, saving you some work
+
+##### What does such a report look like?
+
+Here is an example with no op-to-op tracing. You can instantly see that more time (756us) is spent in between ops (op-to-op gap) than running ops on device (362us)!
+
+Reducing op-to-op gap
+
+![op-to-op gap](images/4.6-op-to-op-gap.png)
+
+There are two main contributors to op-to-op gap: **host time** and **dispatch time**.
+
+* **Host time** is optimized in steps 1-3. If you are already using tracing or are using async mode and have ensured that your python thread is dispatching faster than the device is generating outputs then this has already been minimized.
+* **Dispatch time** is mostly out of your hands, but as an example it is influenced by the number of runtime args a kernel uses.
+    * You can examine the source code for any kernel with high op-to-op latency and see if you can convert some of the runtime args into compile-time args for your use case.
+    * You can fuse multiple ops into a single kernel. Examples where this was well worthwhile in the past include `LayerNorm` and `ScaledDotProductAttentionDecode`.
+
+Typically tracing reduces the op-to-op gap below 6us and as of November 2024 there are roadmap plans to reduce this to zero, so as long as your ops are below this level your opportunities for optimization here are limited.
+
+##### Overall op performance advice
```

Comment:
I added a separate Misc. Performance Optimizations section (4.7) after Performance Analysis. This text is perfect for that.

Reply:
Moved to that section